### PR TITLE
feat(popover): add animation on open and close

### LINKED ISSUE #196: [Popover] Add animation
### Overview

showing, hidding, repositioning should be animated

### Examples in other libraries

Radix UI popover is being animated https://www.radix-ui.com/primitives/docs/components/popover

### Who does this impact? Who is this for?

Users - poorer UX

### PR DESCRIPTION
**Pull request by pawel-twardziak**
Closes #196

allows animation on the popover content while opening/closing it.

Pure CSS animation.

`rdxPopoverContentAttributes` is required to get animations work:

```HTML
<ng-template rdxPopoverContent>
  <div rdxPopoverContentAttributes />
</ng-template>
```

It can get disabled with `[cssAnimation]="false"` on the root directive.

Opening or closing animation can get disabled with `[cssOnShowAnimation]="false"` or `[cssOnCloseAnimation]="false"` on the root directive.

It can get customized with `[cssAnimation]="'custom'"` on the root directive and with the neccessary CSS rules applied to the element having `rdxPopoverContentAttributes` on it:

```HTML
<ng-container rdxPopoverRoot [cssAnimation]="'custom'">
  <ng-template rdxPopoverContent>
    <div class="rdx-custom-animation" rdxPopoverContentAttributes />
  </ng-template>
</ng-container>
```

At some point it looks like the animation is veery fast :) but it is just me who is clicking that fast :D

https://github.com/user-attachments/assets/7f0ac3a1-323d-4f2c-8246-c2f56b0cdba9

### CODE REVIEW

**Comment by pawel-twardziak** [comment]
@pimenovoleg is it ok by you?

**Comment by pimenovoleg** [comment]
It's better to find a way to avoid explicitly specifying styles.

Take a look at how it's done in the Accordion, for example. It only defines CSS variables there.

`@keyframes rdxSlideUpAndFade { }` and other styles are defined outside of the primitives.

**Comment by pimenovoleg** [comment]
I’ve added you to the GitHub group, so now you can work with the repository (branches) without needing to fork it.

**Comment by pawel-twardziak** [comment]
> It's better to find a way to avoid explicitly specifying styles.
> 
> Take a look at how it's done in the Accordion, for example. It only defines CSS variables there.
> 
> `@keyframes rdxSlideUpAndFade { }` and other styles are defined outside of the primitives.

Ok, thanks for the suggestion, I will check there

**Comment by pawel-twardziak** [comment]
> It's better to find a way to avoid explicitly specifying styles.
> 
> Take a look at how it's done in the Accordion, for example. It only defines CSS variables there.
> 
> `@keyframes rdxSlideUpAndFade { }` and other styles are defined outside of the primitives.

**Comment by pawel-twardziak** [comment]
Damn, I clicked `Close with comment` accidently instead of just `Comment` :D

**Comment by pawel-twardziak** [comment]
The greatest challenge there was to handle handle animations havaing the overlay

File: packages/primitives/popover/src/popover-content-attributes.component.ts
Comment (pawel-twardziak) [inline packages/primitives/popover/src/popover-content-attributes.component.ts:13]: @pimenovoleg you mean this stuff is unnecessary? Totally styleless solution, am I right?

Comment (pimenovoleg) [reply packages/primitives/popover/src/popover-content-attributes.component.ts:13]: Yes true, lets be headless :)

File: packages/primitives/popover/src/popover-root.directive.ts
Comment (pawel-twardziak) [inline packages/primitives/popover/src/popover-root.directive.ts:256]: @radix-ng/core @radix-ng/maintainers is this `this.document.dispatchEvent` a standard for this lib? I've copied it from the `tooltip` but I cannot see it being used anywhere else

File: packages/primitives/popover/src/popover-content-attributes.component.ts
Comment (pawel-twardziak) [reply packages/primitives/popover/src/popover-content-attributes.component.ts:13]: Ok, get it. I can't see any room for CSS variables so far, are they usefull for `popover`? Any suggestions?

File: packages/primitives/popover/src/popover-root.directive.ts
Comment (pimenovoleg) [reply packages/primitives/popover/src/popover-root.directive.ts:256]: I think this is copy-pasted from the original React implementation. For Angular, it doesn’t seem necessary.

```tsx
const [open = false, setOpen] = useControllableState({
    prop: openProp,
    defaultProp: defaultOpen,
    onChange: (open) => {
      if (open) {
        providerContext.onOpen();

        // as `onChange` is called within a lifecycle method we
        // avoid dispatching via `dispatchDiscreteCustomEvent`.
        document.dispatchEvent(new CustomEvent(TOOLTIP_OPEN));
      } else {
        providerContext.onClose();
      }
      onOpenChange?.(open);
    },
  });
```

And dont forget about SSR support :)

Comment (pawel-twardziak) [reply packages/primitives/popover/src/popover-root.directive.ts:256]: good point, will improve SSR in another PR :)

File: packages/primitives/popover/src/popover.utils.ts
Comment (artembelik) [inline packages/primitives/popover/src/popover.utils.ts:143]: will this be removed before merging into the main branch?

File: packages/primitives/popover/src/popover-arrow.directive.ts
Comment (artembelik) [inline packages/primitives/popover/src/popover-arrow.directive.ts:98]:
```suggestion
```
or is it overwritten here?

File: packages/primitives/popover/src/popover.utils.ts
Comment (pawel-twardziak) [reply packages/primitives/popover/src/popover.utils.ts:143]: yes

File: packages/primitives/popover/src/popover-arrow.directive.ts
Comment (pawel-twardziak) [reply packages/primitives/popover/src/popover-arrow.directive.ts:98]: yes, leftover, it was filled with a value before. Will remove. Thanks.

Comment (pawel-twardziak) [reply packages/primitives/popover/src/popover-arrow.directive.ts:98]: now I use only top ad left. Right and bottom are set to empty string to make sure they have no value e.g. from outside

File: packages/primitives/popover/src/popover-root.directive.ts
Comment (pawel-twardziak) [inline packages/primitives/popover/src/popover-root.directive.ts:249]: To be fixed: `open` event should get emitted once the open animation is done instead of the current synchronous emission

**Comment by mrfrac** [comment]
> The greatest challenge there was to handle handle animations havaing the CDK overlay.
> 
> I am waiting for your comments guys. Let me know in the comments what and where should I change.

try to follow css styles that provided by original react examples (selector with state attributes). open animation is not a porblem. but closing has some pitfalls (overlay can be detached early then closing annimation finished)

**Comment by pawel-twardziak** [comment]
> > The greatest challenge there was to handle handle animations havaing the CDK overlay.
> > I am waiting for your comments guys. Let me know in the comments what and where should I change.
> 
> try to follow css styles that provided by original react examples (selector with state attributes). open animation is not a porblem. but closing has some pitfalls (overlay can be detached early then closing annimation finished)

Thank you for the suggestion. I looked into the `accordion` directives and I utilized `(animationstart)` and `(animationend)` - they do the job. Now it works perfect. The last thing I need to fix is to emit `onShow` event once the open animation is ended.

File: packages/primitives/popover/src/popover-root.directive.ts
Comment (pawel-twardziak) [reply packages/primitives/popover/src/popover-root.directive.ts]: fixed

**Comment by pawel-twardziak** [comment]
@radix-ng/maintainers @radix-ng/core please take a look at this. Ignore `isRdxPopoverDevMode` - once you accept the PR, I will remove it. @pimenovoleg do not merge it please before I the removal :)

**Comment by pawel-twardziak** [comment]
https://github.com/user-attachments/assets/6d8946b9-57b4-473e-95eb-a849f8397753

File: packages/primitives/popover/src/popover.utils.ts
Comment (pawel-twardziak) [reply packages/primitives/popover/src/popover.utils.ts]: removed

**Comment by pawel-twardziak** [comment]
I have removed that `isRdxPopoverDevMode` util. If all is fine, it is ready to merge.

**Comment by pawel-twardziak** [comment]
I have removed that `isRdxPopoverDevMode` util. If all is fine, it is ready to merge.

File: packages/primitives/popover/src/popover.utils.ts
Comment (pimenovoleg) [inline packages/primitives/popover/src/popover.utils.ts:101]: Could you clarify the logic here?

There's an if condition earlier checking for `[RdxPopoverSide.Top, RdxPopoverSide.Bottom]`, but further on, we're only dealing with `left` & `top`.

Comment (pawel-twardziak) [reply packages/primitives/popover/src/popover.utils.ts:101]: > Could you clarify the logic here?
> 
> There's an if condition earlier checking for `[RdxPopoverSide.Top, RdxPopoverSide.Bottom]`, but further on, we're only dealing with `left` & `top`.

@pimenovoleg yes, sure. For absolute positioninig we can use `left` or `right` for `x` positioning and `top` or `bottom` for `y` positioning. For some cases I used `left` and for the others `right` and the same for `top` and `bottom`. Now. for the sake of simplicity I use only `left` and `top` for absolute positioning. The result is always the same.

**Review by pimenovoleg — APPROVED** [review]

**Comment by pawel-twardziak** [comment]
@pimenovoleg who should squash and merge the PR? I expect it from you but once you added me to the organization, should I proceed that myself? 😸

**Comment by pimenovoleg** [comment]
> @pimenovoleg who should squash and merge the PR? I expect it from you but once you added me to the organization, should I proceed that myself? 😸

The author of the PR should do it :) If you have the necessary permissions, go ahead and click "Merge", using "Squash & Merge".
Since this PR was created from a fork, you might not have the necessary permissions.

**Comment by pawel-twardziak** [comment]
Ok, let's go then... :)

**Pull request merged**
